File: src/Orleans.Security/Authorization/AuthorizationAdmission.cs
using System.Collections.Generic;
using System.Linq;
using System.Reflection;

namespace Orleans.Security.Authorization
{
    internal static class AuthorizationAdmission
    {
        public static bool IsRequired(IGrainCallContext grainCallContext)
        {
            var allowAnonymousAttribute = 
                grainCallContext.InterfaceMethod.GetCustomAttribute<AllowAnonymousAttribute>();
            
            // No authorization required.
            if (allowAnonymousAttribute != null)
            { 
                return false;
            }
            
            var grainMethodAuthorizeData = 
                grainCallContext.InterfaceMethod.GetCustomAttributes<AuthorizeAttribute>(); 
            
            IEnumerable<IAuthorizeData>  grainAuthorizeData =
                    grainCallContext.InterfaceMethod.ReflectedType?.GetCustomAttributes<AuthorizeAttribute>();

                // No authorization required.
            // ReSharper disable once ConvertIfStatementToReturnStatement
            if (grainAuthorizeData != null && !grainAuthorizeData.Any() && !grainMethodAuthorizeData.Any())
            {
                return false;
            }

            return true;
        }
    }
}

File: src/Orleans.Security.FShrap/Authorization/AuthorizationAdmission.fs
namespace Orleans.Security.Authorization

open Orleans
open System.Reflection
open System.Collections.Generic
open System.Linq

type internal AuthorizationAdmission() =
    static member IsRequired(grainCallContext: IGrainCallContext) =
        let allowAnonymousAttribute = grainCallContext.InterfaceMethod.GetCustomAttribute<AllowAnonymousAttribute>()

        if allowAnonymousAttribute <> null then
            false
        else
            let mutable grainAuthorizeData: IEnumerable<IAuthorizeData> = null

            let grainMethodAuthorizeData =
                match grainCallContext.InterfaceMethod with
                | null -> null
                | _ -> grainCallContext.InterfaceMethod.GetCustomAttributes<AuthorizeAttribute>()
            if grainAuthorizeData <> null && not (grainAuthorizeData.Any()) && not (grainMethodAuthorizeData.Any()) then
                false
            else true
